### DISCUSSION TITLE
MCUboot and Zephyr on NXP i.MX RT series

### DISCUSSION THREAD

**Discussion by DerekSnell** [zephyrproject-rtos/zephyr#45252] [Show and tell] [answered]
I am sharing these lab guide steps written in 2021 to help users working with MCUboot and Zephyr.  These are the steps we used on MIMXRT1060-EVK with Zephyr v2.6.0:

> west build -b mimxrt1060_evk -d build-mcuboot bootloader/mcuboot/boot/zephyr
> 
> west flash -d build-mcuboot
> 
> west build -b mimxrt1060_evk -d build-smp-svr-v0 zephyr/samples/subsys/mgmt/mcumgr/smp_svr/ -- -DOVERLAY_CONFIG='overlay-shell.conf' -DCONFIG_MCUBOOT_SIGNATURE_KEY_FILE="bootloader/mcuboot/root-rsa-2048.pem"
> 
> west flash -d build-smp-svr-v0

Use [mcumgr ](https://docs.zephyrproject.org/latest/services/device_mgmt/mcumgr.html)tool to manage new firmware update through UART:

> mcumgr conn add COMPORT type="serial" connstring="dev=COM17,baud=115200,mtu=512"
> 
> mcumgr -c COMPORT image list

The serial terminal will return details like this about the SMP application in MCUboot Slot0:

> Images:
> image=0 slot=0
> version: 0.0.0
> bootable: true
> flags: active confirmed
> hash: 74167b9373435a7a92de33143fbecd888a5b88a23adcc894840e66c2e649023f
> Split status: N/A (0)

The previous steps flashed v0.0.0 of the SMP app in Slot0.  The steps below use mcumgr to update with v1.0.0:

> west build -b mimxrt1060_evk -d build-smp-svr-v1 zephyr/samples/subsys/mgmt/mcumgr/smp_svr/ -- -DOVERLAY_CONFIG='overlay-shell.conf' -DCONFIG_MCUBOOT_SIGNATURE_KEY_FILE=\"bootloader/mcuboot/root-rsa-2048.pem\" -DCONFIG_MCUBOOT_EXTRA_IMGTOOL_ARGS=\"--version=1.0.0\"
> 
> mcumgr -c COMPORT image upload -e build-smp-svr-v1/zephyr/zephyr.signed.bin
> 
> mcumgr -c COMPORT image list

Serial Terminal prints results like this:

> Images:
> image=0 slot=0
> version: 0.0.0
> bootable: true
> flags: active confirmed
> hash: 74167b9373435a7a92de33143fbecd888a5b88a23adcc894840e66c2e649023f
> image=0 slot=1
> version: 1.0.0
> bootable: true
> flags:
> hash: 869fb41491c3262405fcd34626c1a33800d08498642ced5ac693a53f1b15cddb
> Split status: N/A (0)

> mcumgr -c COMPORT image test 869fb41491c3262405fcd34626c1a33800d08498642ced5ac693a53f1b15cddb
> 
> mcumgr -c COMPORT image confirm 869fb41491c3262405fcd34626c1a33800d08498642ced5ac693a53f1b15cddb

**Comment by vishnusudhan** [comment]
Hi @DerekSnell 

I tried the above steps for RT1160 EVK platform,
I could able to flash the MCUBoot & smp-svr-v0 application & able to get output as below,

![Screenshot from 2022-05-24 19-26-28](https://user-images.githubusercontent.com/104553024/170056260-98bb7cdd-abcb-41be-8a17-29fb9ca878ea.png)

After that i try to upload image through mcumgr tool,
1. Added  mcumgr conn add COMPORT type="serial" connstring="dev=COM17,baud=115200,mtu=512"
2.  Isssed command "mcumgr -c COMPORT image list" , I am getting error as ERROR: NMP Timeout.

I have seen  Unable to upload firmware over serial with mcumgr #9509 this thread. Based on that i try to increase the timeout till 20 sec & also try to enable the UART option in the application and getting error as below,


![Screenshot from 2022-05-24 19-35-32](https://user-images.githubusercontent.com/104553024/170057372-19f0505f-bdbe-4413-a6ab-d8a2527bfee0.png)

Suggest is there any steps i am missing out/  need to config the mcuboot/ application to rectify the time out issue?

**Reply by jamesshawver1** [reply]
I know this is on a dead thread, but for anyone else that experiences errors like the one seen here - the issue based upon the image is most likely a device tree issue. It doesn't appear that the uart module has the "chosen" keyword for "uart_mcumgr". I don't remember the exact syntaxy here and zephyr does weird stuff with converting between spaces and commmas and whatnot to underscores - so, I'd guess that there needs to be something like chosen="zephyr,uart_mcumgr" in the device tree within the uart node to get past this compilation error.

**Comment by DerekSnell** [comment]
Thanks @jamesshawver1  for pointing that out.
Specifically using the [MIMXRT1060-EVK devicetree](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/mimxrt1060_evk/mimxrt1060_evk.dts) as an example, it has this chosen node:
```
	chosen {
		zephyr,uart-mcumgr = &lpuart1;
	};

```

**Comment by JPHutchins** [comment]
If you are looking for an alternative to the deprecated mcumgr, we have validated the MIMXRT1060-EVK USB and UDP transports for use with [smpclient](https://pypi.org/project/smpclient/) and maintain HW regression tests here: https://github.com/intercreate/smpclient/tree/main/examples/duts.  I expect that the raw UART would work nicely as well!

**Comment by DerekSnell** [comment, marked as answer]
Marking this example as answered
